homeassistant/components/zamg/weather.py

```diff
@@ -45,53 +46,69 @@
+        "will be removed in a further release of Home Assistant; Your existing configuration "
+        "has been imported into the UI automatically and can be safely removed "
+        "from your configuration.yaml file"
+    )
+
     latitude = config.get(CONF_LATITUDE, hass.config.latitude)
     longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
-
-    station_id = config.get(CONF_STATION_ID) or closest_station(
-        latitude, longitude, hass.config.config_dir
+    station_id = config.get(CONF_STATION_ID) or await hass.async_add_executor_job(
+        closest_station,
+        latitude,
+        longitude,
+        hass.config.config_dir,
     )
-    if station_id not in zamg_stations(hass.config.config_dir):
-        _LOGGER.error(
-            "Configured ZAMG %s (%s) is not a known station",
-            CONF_STATION_ID,
-            station_id,
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data={CONF_STATION_ID: station_id},
         )
-        return
+    )
 
-    probe = ZamgData(station_id=station_id)
-    try:
-        probe.update()
-    except (ValueError, TypeError) as err:
-        _LOGGER.error("Received error from ZAMG: %s", err)
-        return
 
-    add_entities([ZamgWeather(probe, name)], True)
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the ZAMG weather platform."""
+    async_add_entities([ZamgWeather(hass.data[DOMAIN][entry.entry_id])])
 
 
-class ZamgWeather(WeatherEntity):
+class ZamgWeather(CoordinatorEntity, WeatherEntity):
     """Representation of a weather condition."""
 
-    def __init__(self, zamg_data, stationname=None):
+    def __init__(self, coordinator: DataUpdateCoordinator) -> None:
         """Initialise the platform with a data instance and station name."""
-        self.zamg_data = zamg_data
-        self.stationname = stationname
+        super().__init__(coordinator)
+        self._attr_unique_id = f"{coordinator.data.get(CONF_STATION_ID)}"
+        self._attr_name = f"{coordinator.data.get('station_name')}"
+        self._attr_device_info = DeviceInfo(
+            entry_type=DeviceEntryType.SERVICE,
+            identifiers={(DOMAIN, coordinator.data.get(CONF_STATION_ID))},
+            manufacturer=ATTRIBUTION,
+            configuration_url=MANUFACTURER_URL,
+            name=coordinator.name,
+        )
+
+    @property
+    def device_info(self):
+        """Return the device info."""
+        return self._attr_device_info
 
     @property
     def name(self):
         """Return the name of the sensor."""
-        return (
-            self.stationname
-            or f"ZAMG {self.zamg_data.data.get('Name') or '(unknown station)'}"
-        )
+        return f"ZAMG {self.coordinator.data.get('station_name')}"
```

Comment:
These are already handled by the base class / `_attr_`-attribute.

Reply:
done